Show error and warning indicators in project panel items

nilskch opened the pull request:
Closes #5016

Release Notes:

- Add setting to display error and warning indicators in project panel items.

https://github.com/user-attachments/assets/8f8031e6-ca47-42bf-a7eb-718eb1067f36

nilskch committed:
show errors in project panel

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -237,6 +238,12 @@ impl ProjectPanel {
                 project::Event::ActivateProjectPanel => {
                     cx.emit(PanelEvent::Activate);
                 }
+                project::Event::DiskBasedDiagnosticsFinished { .. }
+                | project::Event::DiagnosticsUpdated { .. } => {
+                    this.update_diagnostics(cx);
+                    this.update_visible_entries(None, cx);
+                    cx.notify();
+                }
```
We could add a property in the settings to opt in or out of this feature and could not update the diagnostics in case someone does not want the feature.

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -433,6 +441,32 @@ impl ProjectPanel {
+            .read(cx)
+            .diagnostic_summaries(false, cx)
+            .filter_map(|(path, _, diagnostic_summary)| {
+                if diagnostic_summary.error_count > 0 {
+                    Some((path, Color::Error))
+                } else {
+                    None
+                }
```
I thought about doing this:
```rust
if diagnostic_summary.error_count > 0 {
    Some((path, Color::Error))
} else if else if diagnostic_summary.warning_count > 0 {
    Some((path, Color::Warning))
} else {
    None
}
```
I didn't like that and removed it. We could make this configurable if you want?

nilskch reviewed (commented):
Hey, I love Zed and everything you do for the community! I was missing this feature a lot and many people asked for it and I wanted to give it a try. I hope this is something that you would want for Zed, let me know if I should make any changes.

nilskch committed:
color root directory as well

nilskch committed:
fix clippy

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -82,6 +82,7 @@ pub struct ProjectPanel {
     show_scrollbar: bool,
     scrollbar_drag_thumb_offset: Rc<Cell<Option<f32>>>,
     hide_scrollbar_task: Option<Task<()>>,
+    diagnostic_error_colors: HashMap<PathBuf, Color>,
```
if we do want to support showing warnings in the project panel, this should we a HashMap. If we only want to show errors we can make this a HashSet and not store the colors.

nilskch commented:
Okay I played around a little bit and I have a concrete suggestion:
Let's add a `"show_diagnostic_errors"` bool property in the `"project_panel"` object in the `settings.json`. I would use a default value of `true`, but happy to make it `false` if you like it better. 

I would not show warnings in the project panel. I think this causes too much unnecessary noise. The status `GitFileStatus::Modified` paints the items with the color `Color::Modified` which is very similar to `Color::Warning` (at least in the theme `One Light`) and this is very confusing.

If people want to display warnings as well, I am happy to make a follow-up PR but I honestly do not think someone wants this.

I will implement this and add the docs next week if you think this would be a good add.

nilskch committed:
add configuration for diagnostic errors

nilskch commented:
I implemented it the way I suggested in the last comment. Looking forward to your feedback on this!

https://github.com/user-attachments/assets/2fe4465d-afbb-4edb-abfe-e7d33991b0a5

nilskch committed:
some optimizations

nilskch committed:
nit

nilskch committed:
docs nit

nilskch committed:
final cleanup

nilskch commented:
@mrnugget can you please take a look at this PR ? Tagging you because of the time zone^^

mrnugget commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -82,6 +82,7 @@ pub struct ProjectPanel {
     show_scrollbar: bool,
     scrollbar_drag_thumb_offset: Rc<Cell<Option<f32>>>,
     hide_scrollbar_task: Option<Task<()>>,
+    diagnostic_error_paths: HashSet<PathBuf>,
```
I think you need something like `HashMap<WorktreeId, HashSet<PathBuf>`, otherwise you'll have global diagnostic errors, even if they appear in a different worktree.

Example:
![screenshot-2024-09-24-15 00 32@2x](https://github.com/user-attachments/assets/b6d6ea08-94cf-4fee-b41c-4ced45a5aade)

Here I have an error in the rust-proj, but the go-proj is also marked red.

(cc @bennetbo who found this bug)

nilskch replied:
I didn't know you can do this in Zed, thanks for flagging!

mrnugget replied:
Takes a while to digest that and then you see it everywhere :)

mrnugget reviewed (commented):
Thank you! Left a bunch of comments :)

bennetbo commented:
> If people want to display warnings as well, I am happy to make a follow-up PR but I honestly do not think someone wants this.\

I'd like to have this and would prefer to add this in this PR as well, so that we do not have to break the settings right after this is released.
Maybe we could have a setting like (not sure about the names of the enum variants):
`"show_diagnostics": "off/all/errors"`

> I would not show warnings in the project panel. I think this causes too much unnecessary noise. The status GitFileStatus::Modified paints the items with the color Color::Modified which is very similar to Color::Warning (at least in the theme One Light) and this is very confusing.

Let's add the warnings as an option and we will figure out how to make it "pretty" later.

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -279,9 +288,14 @@ impl ProjectPanel {
+            cx.observe_global::<SettingsStore>(move |this, cx| {
                 let new_settings = *ProjectPanelSettings::get_global(cx);
                 if project_panel_settings != new_settings {
+                    if new_settings.show_diagnostic_errors
+                        && !project_panel_settings.show_diagnostic_errors
+                    {
+                        this.update_diagnostics(cx);
+                    }
```
@mrnugget my thought was that `self.diagnostic_error_paths` is not used and updated once you enable the setting again and therefore it's not necessary to reset it. On the other hand this wastes memory, I will reset it.

nilskch committed:
Update crates/project_panel/src/project_panel_settings.rs

Co-authored-by: Thorsten Ball <mrnugget@gmail.com>

nilskch committed:
add warnings to project panel

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -82,7 +85,7 @@ pub struct ProjectPanel {
     scrollbar_drag_thumb_offset: Rc<Cell<Option<f32>>>,
     hide_scrollbar_task: Option<Task<()>>,
-    diagnostic_error_paths: HashSet<PathBuf>,
+    diagnostics: HashMap<(WorktreeId, PathBuf), DiagnosticSeverity>,
```
@mrnugget I went for a tuple to avoid chaining HashMaps if that is ok for you?

mrnugget replied:
Absolutely, good idea.

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -291,10 +295,12 @@ impl ProjectPanel {
+                    if new_settings.show_diagnostics != project_panel_settings.show_diagnostics {
+                        if new_settings.show_diagnostics != ShowDiagnostics::Off {
+                            this.update_diagnostics(cx);
+                        } else {
+                            this.diagnostics = Default::default();
+                        }
```
Resetting this now

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -449,30 +455,53 @@ impl ProjectPanel {
-                    .collect::<Vec<_>>()
+            .for_each(|(project_path, diagnostic_severity)| {
+                let mut path_buffer = PathBuf::new();
+                self.update_highest_diagnostic_severity(
+                    &project_path,
+                    path_buffer.clone(),
+                    diagnostic_severity,
+                );
```
This here is necessary to also include the root directory of the worktree. We need the empty buffer in the HashMap as well.

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -449,30 +455,53 @@ impl ProjectPanel {
+        self.diagnostics
+            .entry((project_path.worktree_id, path_buffer.clone()))
+            .and_modify(|max_diagnostic_severity| {
+                *max_diagnostic_severity =
+                    std::cmp::min(*max_diagnostic_severity, diagnostic_severity);
```
It's a little bit confusing, but `DiagnosticSeverity::ERROR < DiagnosticSeverity::WARNING` and I think errors should overshadow warnings here.

nilskch replied:
EDIT: renamed to `strongest_diagnostic_severity`

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -2242,8 +2281,10 @@ impl ProjectPanel {
+            details.diagnostics_color.unwrap()
         } else {
             entry_git_aware_label_color(details.git_status, details.is_ignored, is_marked)
         };
```
Let me know if you know a nicer way to write this :) I don't really like the `unwrap()`

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -449,30 +455,53 @@ impl ProjectPanel {
-            .project
+        let show_warnings =
+            ProjectPanelSettings::get_global(cx).show_diagnostics == ShowDiagnostics::All;
+        self.diagnostics = Default::default();
+        self.project
```
is it better here to store everything in a local HashMap first and assign the local variable at the end to `self.diagnostics`?

mrnugget replied:
Yeah, I think that's how I'd do it, would make the code nicer, IMHO

nilskch reviewed (commented):
@mrnugget @bennetbo thanks for the feedback! I think I addressed your comments now. Let me know how you want the settings to look like and what defaults we should use.

nilskch committed:
rename var

nilskch committed:
cleanup

nilskch committed:
remove log dependency again

nilskch commented:
Hey so this is how it looks like with warnings and the settings as @bennetbo suggested:

https://github.com/user-attachments/assets/4d866e25-b65c-4304-ba38-f6e0fcf0170f

PS: I changed my GPG key last night and apparently you should not remove the old public key from your profile, otherwise old commits show up as unverified. It's still me, just wiser.

mrnugget commented:
One thing that I'm now wondering about: how will a modified file look different from a file with warnings? Right now, modified files look like this: 
![screenshot-2024-09-25-10 15 52@2x](https://github.com/user-attachments/assets/fa1b752c-4ca9-469f-8337-0bf069a14b0d)

That's also the has-a-warning style, no?

nilskch commented:
@mrnugget in the last comment I posted a video where you can see the difference when I changed the settings. In the theme "One Light" the warning color is a lot brighter than the modified color. Tbh, I am not the biggest fan of having both, warnings and git modified. 

> Let's add the warnings as an option and we will figure out how to make it "pretty" later.

Not sure how much effort it is to make it "pretty", but maybe we should do that before we introduce this feature? The contrast of the warnings color is very bad for the selected file (see video).

danilo-leal commented:
I'd avoid using only colors for these statuses—different themes will have different color qualities, and relying only on them can be very tricky. We can't guarantee that it will look great in all themes, and folks with any visual impairments can have an even harder time! 

That said, I'd consider adding the same icons we display on the status bar for errors and warnings and simple dot indicators for the Git-related statuses (added & modified) to the project item. I'd also add tooltips to them.

![image](https://github.com/user-attachments/assets/d52ac53e-9f55-42f6-878f-80d3bbda81dd)

zackangelo commented:
Thank you for working on this! It's the only thing keeping me from using Zed as my main editor at this point. 🙂

iamnbutler commented:
Hey @nilskch! Thanks for contributing. 

I'm looking into some design considerations around this today, I'll come back and add more detail shortly – just adding a quick note so you aren't waiting around with no feedback.

nilskch commented:
@iamnbutler sounds good to me, thanks!

nilskch commented:
@iamnbutler @danilo-leal this is how @danilo-leal suggestion looks like:

<img width="1127" alt="with_dots" src="https://github.com/user-attachments/assets/c3f1599a-c582-4c5d-bc67-6b77f5659f05">

And just for fun without the dots for a git status:

<img width="1127" alt="without_dots" src="https://github.com/user-attachments/assets/bf5262fe-5595-48f6-9191-a5b2666eae2b">

nilskch committed:
test design with additional icons

nilskch committed:
merge main

nilskch committed:
cleanup

CharlesChen0823 commented:
personly, I like the icon ends with alignment.

CharlesChen0823 commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -2487,6 +2497,15 @@ impl ProjectPanel {
                         }
                         .ml_1(),
                     )
+                    .child(if let Some(icon_name) = extra_icon_name {
```
you can using `end_slot` in here

nilskch commented:
@CharlesChen0823 very nice idea!

<img width="889" alt="end_slots" src="https://github.com/user-attachments/assets/9fd01a50-2a66-4c52-a351-3af88541f891">

mrnugget commented:
I think the dots in addition are a bit much, but with only the icons, I'd say we can merge this.

(If you want: I'd be interested to see what it would look like if the opacity of the icons color is turned to 70% if they are warning/error colored, because those red folders really stick out)

mrnugget commented on assets/settings/default.json:
```diff
@@ -359,7 +359,11 @@
-    }
+    },
+    /// Which files containing diagnostic errors/warnings to mark in the project panel.
+    ///
```
Can you document the other possible values here?

mrnugget commented on crates/editor/src/items.rs:
```diff
@@ -1481,6 +1482,43 @@ pub fn entry_label_color(selected: bool) -> Color {
+    git_status: Option<GitFileStatus>,
+    ignored: bool,
+) -> Option<IconName> {
+    if let Some(diagnostic_severity) = diagnostic_severity {
+        if *diagnostic_severity == DiagnosticSeverity::ERROR {
+            return Some(IconName::XCircle);
+        }
+        if *diagnostic_severity == DiagnosticSeverity::WARNING {
+            return Some(IconName::Warning);
+        }
+    }
+    if !ignored && git_status.is_some() {
+        Some(IconName::Dot)
+    } else {
+        None
+    }
```
Nitpick, but I'd use a `match` statement:

```suggestion
    match diagnostic_severity {
        Some(&DiagnosticSeverity::ERROR) => Some(IconName::XCircle),
        Some(&DiagnosticSeverity::WARNING) => Some(IconName::Warning),
        _ => {
            if !ignored && git_status.is_some() {
                Some(IconName::Dot)
            } else {
                None
            }
        }
    }
```

mrnugget commented on crates/editor/src/items.rs:
```diff
@@ -1481,6 +1482,43 @@ pub fn entry_label_color(selected: bool) -> Color {
+    }
+}
+
+pub fn entry_diagnostic_and_git_aware_label_color(
```
I'd actually change these two functions to be a single function that returns `(Color, Option<IconName>)`

mrnugget commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -279,9 +295,16 @@ impl ProjectPanel {
                 let new_settings = *ProjectPanelSettings::get_global(cx);
                 if project_panel_settings != new_settings {
+                    if new_settings.show_diagnostics != project_panel_settings.show_diagnostics {
+                        if new_settings.show_diagnostics != ShowDiagnostics::Off {
+                            this.update_diagnostics(cx);
+                        } else {
+                            this.diagnostics = Default::default();
+                        }
```
Would be nice to restructure the code so you could always call `this.update_diagnostics(cx)` in case the value changed. Since `update_diagnostics` already resets the diagnostics, the rest of it would only need to be restructured so it only adds diagnostics based on the settings (which it already kinda does by using `show_warnings`)

nilskch commented:
I agree with the dots. I change the opacity to 70% this evening (CET) and see how it looks like :)

Also happy to implement anything that @iamnbutler proposes!

What do you think about the settings? They are hard to change once we merge and we should get them right. Theoretically there are more diagnostic severities (`INFORMATION` and `HINT` and maybe more in the future?) and I think we should design the settings in a way that they are extendable. Also, the current design doesn't allow for just showing warnings and not errors (I do not want this but still wanted to mention it). We could add an extra enum variant for that but the number of variants would explode with the number of severity types if you want to support every combination.

nilskch commented on assets/settings/default.json:
```diff
@@ -359,7 +359,11 @@
+    /// Which files containing diagnostic errors/warnings to mark in the project panel.
+    ///
+    /// Default: "errors"
+    "show_diagnostics": "errors"
```
And what default setting do you prefer? With the Icons I actually prefer "all"

jansol commented:
What if the dots are placed right after (or even on the left side of) the file name and only the icons at the right edge? To me they are two very orthogonal features and I think I'd want to have both most of the time.

nilskch committed:
implement feedback

nilskch committed:
fix default value in docs

nilskch committed:
cleanup

nilskch commented:
Thanks for the feedback everyone! I applied the changes from the review and cleaned up a little. I am happy with the current status, but let me know if I should make additional changes!

> What if the dots are placed right after (or even on the left side of) the file name and only the icons at the right edge? To me they are two very orthogonal features and I think I'd want to have both most of the time.

@jansol I think the dots could be a very nice feature but since this is somewhat orthogonal I'd say we discuss this in a separate issue to not overcomplicate things here?

danilo-leal commented:
Hey @nilskch, thanks so much for the continued work here! Just to give you an update, @iamnbutler & I have looked into this change yesterday and tried to figure out the design taking into consideration other project panel-related changes going on at the moment (particularly horizontal scrolling) that would affect what we're doing here as those add new complexities.

So, the approach we came up with that we've enjoyed the most was this one:

<img width="500" alt="Screenshot 2024-09-27 at 9 12 31 AM" src="https://github.com/user-attachments/assets/dd0a235a-e0be-4730-b9bb-e5cf8be6bce6">

Visually speaking, it's very different than what we're playing with here, but it has some pros. It works with horizontal scrolling and doesn't overloads the project panel items with too many colors—we want to be conscious about the growing visual complexity in Zed. It also might have some yet unanswered questions, like:
- How do we display, on a folder, that there's a file with diagnostic inside it? (The `cloudflare` example directory in the picture doesn't feel like the best solution yet)
- How do we move on with this approach if file icons are disabled? Maybe what we have in the second picture suffices? Or would it be confusing because of how an X icon could be confused with a close action?

That said, we might take a bit more time to go through these, before pushing through this PR, and then give an update here with clearer designs. We might also discuss horizontal scrolling and its implications in taking on this convo! Thanks for bearing with us 😊

nilskch commented:
Hey, thanks for taking the time on this! I completely understand your points. This feature is quite important to me (and I think for most other VS Code transitioners as well), so I am happy to spend more time on this to make it happen. Feel free to use me as a resource to implement different designs to figure out what works best!

nilskch commented:
Hey @Angelk90, sorry for the late reply! The latest version on this branch is the most left version of the image in [this comment](https://github.com/zed-industries/zed/pull/18182#issuecomment-2377645115).

I really like the design, but as far as I understood, the Zed team is working on horizontal scrolling in the project panel at the moment and I do not think the A/M icons would work with that. I think it is best to wait for feedback from the Zed team before implementing more designs :)

WeetHet commented:
> Hey @nilskch, thanks so much for the continued work here! Just to give you an update, @iamnbutler & I have looked into this change yesterday and tried to figure out the design taking into consideration other project panel-related changes going on at the moment (particularly horizontal scrolling) that would affect what we're doing here as those add new complexities.
> 
> So, the approach we came up with that we've enjoyed the most was this one:
> 
> <img alt="Screenshot 2024-09-27 at 9 12 31 AM" width="500" src="https://private-user-images.githubusercontent.com/67129314/371431720-dd0a235a-e0be-4730-b9bb-e5cf8be6bce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjgyNzY0MTksIm5iZiI6MTcyODI3NjExOSwicGF0aCI6Ii82NzEyOTMxNC8zNzE0MzE3MjAtZGQwYTIzNWEtZTBiZS00NzMwLWI5YmItZTVjZjhiZTZiY2U2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDA3VDA0NDE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4NDk4NWM2OWQ5MjAwN2Q0MWFjNDgzZTA2ZjM5YmZmYzY1ZmI5ODI4ZTY3ZWQxNmNiNWNjMTMwZGUzMjY3ZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ll8sGdxokdt0KxTZj53rmX7MJHfjy-Ic8ot1nvbkTIo">
> Visually speaking, it's very different than what we're playing with here, but it has some pros. It works with horizontal scrolling and doesn't overloads the project panel items with too many colors—we want to be conscious about the growing visual complexity in Zed. It also might have some yet unanswered questions, like:
> 
> * How do we display, on a folder, that there's a file with diagnostic inside it? (The `cloudflare` example directory in the picture doesn't feel like the best solution yet)
> * How do we move on with this approach if file icons are disabled? Maybe what we have in the second picture suffices? Or would it be confusing because of how an X icon could be confused with a close action?
> 
> That said, we might take a bit more time to go through these, before pushing through this PR, and then give an update here with clearer designs. We might also discuss horizontal scrolling and its implications in taking on this convo! Thanks for bearing with us 😊

I think this looks incredible to be honest. To address your concerns, no, the version without file icons looks great and I would never expect that to be a close action. As per the directory diagnostics, while the suggested approach might not be the best, it already is much better than either the previous approach used in this PR or the current situation where you can't see anything in the project panel. In my opinion, merging this under an experimental settings would be nice to have, after the suggested approach is implemented

nilskch commented:
Hey, I totally agree with @WeetHet, the design looks great! I dont know what the status on horizontal scrolling is, but do you have any new design ideas? Or can I implement your solution?

WeetHet commented:
> Hey, I totally agree with @WeetHet, the design looks great! I dont know what the status on horizontal scrolling is, but do you have any new design ideas? Or can I implement your solution?

I'd say just implement the @danilo-leal's design, as it does not seem to require any integration with horizontal scrolling

nilskch committed:
merge main

nilskch commented:
Hey, what do you think about this? I tried to implement @danilo-leal 's suggestion as close as possible.

(It's a little bit slow because I am using `cargo run` here)


https://github.com/user-attachments/assets/99f4a5f4-170f-4041-9f25-60262576d0a9

nilskch committed:
implement new design

WeetHet commented:
> Hey, what do you think about this? I tried to implement @danilo-leal 's suggestion as close as possible.
> 
> (It's a little bit slow because I am using `cargo run` here)
> 
>  demo.mov

Amazing work! Just a couple of small nitpicks:
- Maybe make the icons a bit smaller so they don't overlap with the file icons that much, especially, when folder icons are disabled
- When the file icons are disabled, I would prefer for the files to still be aligned, so maybe move the files with the errors a bit to the right, as it is done in the original proposal

nilskch committed:
make warning icon smaller

nilskch commented:
Forgot to record folder icons with warnings...

https://github.com/user-attachments/assets/647ba041-9988-4f2e-8438-efd2b8afd940

> Maybe make the icons a bit smaller so they don't overlap with the file icons that much, especially, when folder icons are disabled 

I'm currently using the already existing `DecoratedIcon` and the X already existed with this size. I would need to change the lib to allow a different size for the X, but I made the triangle icon a bit smaller (i created that one).

> When the file icons are disabled, I would prefer for the files to still be aligned, so maybe move the files with the errors a bit to the right, as it is done in the original proposal

Not sure what you mean, I just swap the file icon (or the empty placeholder of the file icon) with an X or a triangle. Can you explain this a little more?

WeetHet commented:
> the empty placeholder of the file icon

For me the empty placeholders look weird, I'd look into removing those, because they just add extra spacing to the left. I'd propose to replace the current behaviour with `"file_icons": "empty"` and add a `"file_icons": "none"` option that would completely get rid of that space. If a file has errors of warnings, it would move that file to the right and add a respective icons on the left. That behaviour matches @danilo-leal's renders a bit better, and in my mind looks better too

Also, it seems like in the renders, the opened folders don't have an indicator. I might only imagine it's intentional, as you can see the error below anyways, and an angle icon with a cross pointing down doesn't really look great

nilskch committed:
cleanuo

WeetHet commented:
> * The upper folders also don't have the warning, why?
> * But the `.cloudflare` folder has the warning, why?

It seems to me that only the folders that are not open show a warning/error sign, and to me it seems reasonable because it’s enough to inspect the contents instead

WeetHet commented:
> @WeetHet : As you say it could be fine. In the current implementation it is always displayed.
> 
> But if the folder is closed and there is a warning and a yellow triangle, what do you show in the folder the warning?

You must mean an error and a warning, because a yellow triangle signalises a warning. In case you have both an error and a warning, an error always takes precedence, as it is only reasonable to first make sure your code compiles

nilskch commented:
> It seems to me that only the folders that are not open show a warning/error sign, and to me it seems reasonable because it’s enough to inspect the contents instead

Ah, good point! I didn't get that. But I actually like it better to show errors/warnings on open folders as well. Sometimes there is a lot of stuff inside a folder and the files that contain errors are not visible in the project panel and it helps if all parent folders show the diagnostics (this is just my opinion, totally understand if you prefer the opposite).

I think we gave enough input and it's best to wait for feedback from the maintainers here. I'm also happy to pair program on this PR if the Zed team thinks it makes sense (I read in the [contributing guidelines](https://github.com/zed-industries/zed/blob/main/CONTRIBUTING.md#pair-programming) that you sometimes do this).

nilskch committed:
merge main

nilskch committed:
merge main

nilskch commented:
Hey, fantastic job with the indents and the horizontal scrolling in the project panel, I love it! Are the diagnostics something that you would want to add to the project panel as well? Feedback from the Zed team here is highly appreciated :)

nervenes commented:
> I really like the design, but as far as I understood, the Zed team is working on horizontal scrolling in the project panel at the moment and I do not think the A/M icons would work with that. I think it is best to wait for feedback from the Zed team before implementing more designs :)

Hey, i've just created a PR for git symbols in the project panel (and editor tabs), might want to check it out: https://github.com/zed-industries/zed/pull/20134

mrnugget committed:
project panel: Make diagnostic error status overwrite git status

Co-authored by: Danilo Leal <danilo@zed.dev>
Co-authored by: Peter Tripp <peter@zed.dev>

danilo-leal committed:
Merge branch 'main' into pr/18182

danilo-leal committed:
Checkpoint: Create a dedicated icon for panel diagnostic display

danilo-leal committed:
Merge branch 'main' into pr/18182

nilskch committed:
Merge branch 'zed-industries:main' into feat/highlight-errors-in-project-panel

danilo-leal committed:
Merge branch 'main' into pr/18182

danilo-leal committed:
Use the new `DecoratedIcon` component

danilo-leal committed:
Organize colors used for each state and scenario

danilo-leal committed:
Remove unused SVGs

danilo-leal committed:
Iterate on the setting's writing

danilo-leal committed:
Iterate on the triangle knockout icon

danilo-leal reviewed (approved):
I'll merge this PR for now; feeling good about it! Thanks so much @nilskch and everyone else for the continued effort & contribution! We'll tackle any upcoming feedback iteratively on future PRs.

nilskch commented on crates/project_panel/src/project_panel.rs:
```diff
@@ -2416,8 +2418,14 @@ impl ProjectPanel {
 
-                    let filename_text_color =
-                        entry_git_aware_label_color(status, entry.is_ignored, is_marked);
+                    let filename_text_color = if entry.kind.is_file()
+                        && diagnostic_severity
+                            .map_or(false, |severity| severity == DiagnosticSeverity::ERROR)
+                    {
+                        Color::Error
+                    } else {
+                        entry_git_aware_label_color(status, entry.is_ignored, is_marked)
+                    };
```
Yes sounds good :) Thank you so much @danilo-leal, the design looks fantastic!

nervenes commented:
@danilo-leal i see, i know this is just an initial implementation but i think there are some things that can be cleared up, im trying to think from the average users perspective, but y'all know the best :>

if i were new, and the project panel showed me warning and error indicators on the file icon for diagnostics, and the file name was yellow (modified) or red (conflict) for git, and i solved the warning/error diagnostic problem but the filename is still colored yellow/red (for git) would confuse me, and vice versa if i solved a git problem and the yellow/red filename color went away while the warning/error indicator for diagnostics persisted. what's even more confusing now is that diagnostics coloring and git coloring for filenames take mixed priority, error diagnostics take highest priority, git takes second priority than warning diagnostics, and i know this because i've followed through the development of this but will the regular user know this?

if this is the path that is wanted, then maybe it would be better to either pick diagnostics or git (not both) for coloring? so i'd imagine with coloring mode to be set to diagnostics, it would color filenames either red or yellow, with no conflicting git colors, and git symbols (when/if merged) will be used to infer git information. or the opposite, that git coloring is enabled and you'd only infer diagnostics from the indicators (but the problem here is that the indicators are colored and small, so it still may not be clear to the user what it's actually for). i would personally recommend the former.

danilo-leal commented:
@nervenes I just pushed a PR earlier today that I think goes along your latter suggestion https://github.com/zed-industries/zed/pull/20600. I hear you about the icons being potentially not clear, though. I guess we could solve it via a tooltip. Like, if you hover the file with the X error icon, you could see something like "There's an error" or a better label.

nervenes commented:
additionally, maybe clicking on the tooltip could take you to the particular error/warning in the diagnostic panel? also, is there any other place i could have a discussion with you, maybe any of the channels in the discord channel? i'd like to ask some questions about the git symbols, or should i open up an empty pr?

hawkinsonb commented:
@danilo-leal @nervenes Why not strike-through the filename for a git removal and use a darker red color if you need a color indicator? That way it very clearly separates it from diagnostics red, while adding a visual cue that a file was removed. This would accomplish the separation of functionality you are looking for while making the UX much nicer and explicit.

danilo-leal commented:
@nervenes interesting idea! Hit me up on Zed's Discord. You can fine me there!

@hawkinsonb curious to explore that, yeah. Thanks for the suggestion!
